src/basic/cgroup-util.c

```diff
@@ -95,35 +95,44 @@ int cg_enumerate_processes(const char *controller, const char *path, FILE **ret)
         return cg_enumerate_items(controller, path, ret, "cgroup.procs");
 }
 
-int cg_read_pid(FILE *f, pid_t *ret) {
+int cg_read_pid(FILE *f, pid_t *ret, CGroupFlags flags) {
         unsigned long ul;
 
         /* Note that the cgroup.procs might contain duplicates! See cgroups.txt for details. */
 
         assert(f);
         assert(ret);
 
-        errno = 0;
-        if (fscanf(f, "%lu", &ul) != 1) {
+        for (;;) {
+                errno = 0;
+                if (fscanf(f, "%lu", &ul) != 1) {
 
-                if (feof(f)) {
-                        *ret = 0;
-                        return 0;
+                        if (feof(f)) {
+                                *ret = 0;
+                                return 0;
+                        }
+
+                        return errno_or_else(EIO);
                 }
 
-                return errno_or_else(EIO);
-        }
+                if (ul > PID_T_MAX)
+                        return -EIO;
 
-        if (ul <= 0)
-                return -EIO;
-        if (ul > PID_T_MAX)
-                return -EIO;
+                /* In some circumstances (e.g. WSL), cgroups might contain unmappable PIDs from other
+                 * contexts. These show up as zeros, and depending on the caller, can either be plain
+                 * skipped over, or refused and returned as an error. */
+                if (ul <= 0) {
+                        if (FLAGS_SET(flags, CGROUP_REFUSE_UNMAPPED))
+                                return -EREMOTE;
```

Comment:
hmm, i don't like that this is returned as an error, because this isn#t really an errror, but just a special case, and it might very well happen that cgroup.procs shows a zero process first, and a non-zero process next, so if we trigger an error here, we'd abort the enumeration, and not see the other items anymore.

Hence, I think we should encode this differently here. maybe something like this:

```c
if (FLAGS_SET(flags, CGROUP_REFUSE_UNMAPPED)) {
        *ret = 0;
        return 1;
}
```

i.e. simply actually return zero as pid here, and > 0 as return value.

Reply:
So in a sense revert to one or two previous instances, where the flag was still called CGROUP_INCLUDE_UNMAPPED, and where it toggled between returning them as pid vs. just skipping over them?

Reply:
IOW: 0 is not accepted by pud_is_valid and friends, i.e. very much unexpected, so treating it as if it's a valid value seems wrong. OTOH a recognizable errno still allows the caller to handle it gracefully, but the intention is clearer.

Reply:
I have to agree that it's cleaner to handle this if a pid_t can never be zero here.
For the COUNT_*_PROCESSES case I just ammended, if the returned pid_t could be zero, it'd be relying on the pid_t having an invalid value that could cause unknown effects if not accounted for everywhere.

Reply:
Ah, i didn't see that. I guess cg_read_pidref() should get some special handling to deal with zero pids and return an empty pidref instead.

the thing is that this is an enumeration api and tejun is very clear that people should not rely on enumeration order. and if we see a zero pid we should continue enumerating, but returning an error makes this a bit icky.

Reply:
I think when e.g. enumerating through dirs we often ignore a bunch of errors, and this is not so different? We can move the comment about EREMOTE onto the beginning of the function and that's enough?

Reply:
Right now, when we recursively iterate through dirs for some purposes we indeed sometimes react gracefully to some errors, but it's not the errors from readdir() we handle gracefully. instead, it's the errors of the stuff we then execute on the individual dir entries. i.e. failures when reading the next dir entry are generally fatal to us, failures what to then use it for are handled gracefully.

cg_read_pid() here is just a wrapper around readdir(), hence I am pretty sure the least surprising behaviour here would be not to consider a zero PID an error (because it really is not an error), but just as a special case. A case special enough to filter it out by default, but still just a special case, not an error.

Reply:
I have no super strong opinion on either approach, so I'm happy to switch it back to the previous approach of it returning the zeros.
It does need to set a flag for it to return zeros, so there should be no big surprises looming anywhere.

Reply:
I think the approach with just returning 0 seems better. The caller needs to handle the case explicitly, and that's fine.